Improve Creature DBC Handling

HelloKitty opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Remove hardcoded literal flags check for CreatureModelData mount capability.
-  Check for CreatureTypeEntry flags for calculating Creature::CanGiveExperience and remove IsCritter check (defined in DBC along with gascloud and non-combat pets)

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #  (insert issue tracker number)


**Tests performed:** (Does it build, tested in-game, etc.)

Can still gain experience from normal creatures. Cannot get experience from critters.

**Known issues and TODO list:** (add/remove lines as needed)

None

<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

HelloKitty committed:
Core/DBC: Add CreatureModelDataFlags to remove hardcoded literals

HelloKitty committed:
Core/DBC: Use CreatureType flags in Creature::CanGiveExperience calculation

HelloKitty commented:
The build failure is not from my commit, it's from: ee5eace

funjoker commented:
Please revert to your commit and use rebase

HelloKitty commented:
I tried running a rebase but I must have done it wrong. I'm unfamiliar with this area of git. If I keep trying to fix it then it'll probably get worse, like the other PR I opened. I'm sorry.

HelloKitty commented:
Ok, small chance I may have fixed it.

Shauren commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -3245,8 +3245,12 @@ void Creature::ClearTextRepeatGroup(uint8 textGroup)
-        && !IsPet()
+    CreatureTypeEntry const* creatureType = sCreatureTypeStore.AssertEntry(GetCreatureType());
+
+    if (creatureType->HasFlag(CreatureTypeEntryFlags::CREATURE_TYPE_ENTRY_FLAGS_NO_EXPERIENCE))
```
You should revert this change, this flag has a different meaning, its a "do not include in tab targeting"

HelloKitty replied:
Ohhh ok, that actually makes a lot of sense now. The information I was basing it on claimed it was something else.

HelloKitty replied:
Partially reverted that commit now.

HelloKitty committed:
Core/DBC: Fixed CreatureType Flags Misinterpretation

Meaning of this was wrong, wowdev wiki was incorrect. Partially reverts
0d61e653f4841d5c9745639a6bdf7487b4d082fe

Shauren commented on src/server/shared/DataStores/DBCStructure.h:
```diff
@@ -522,12 +529,19 @@ struct CreatureSpellDataEntry
     //char*   Name[16];                                     // 1-16     name
                                                             // 17       string flags
-    //uint32    no_expirience;                              // 18 no exp? critters, non-combat pets, gas cloud.
+    uint32    Flags;                                        // 18 Critters, non-combat pets, gas cloud. (called Flags in struct)
```
There is no reason to load this field anymore

HelloKitty replied:
True, but why is so much effort put into not loading parts of DBC? Their memory footprint seems to be nearly nothing.

Shauren replied:
I have no idea

Shauren commented on src/server/game/Entities/Unit/Unit.cpp:
```diff
@@ -8902,7 +8902,7 @@ bool Unit::IsInDisallowedMountForm() const
     ChrRacesEntry const* race = sChrRacesStore.LookupEntry(displayExtra->Race);
 
-    if (model && !(model->Flags & 0x80))
+    if (model && !(model->HasFlag(CreatureModelDataFlags::CREATURE_MODEL_DATA_FLAGS_CAN_MOUNT)))
```
Codestyle remark: dont prefix regular enums with `Type::`, use that only for `enum class` where its required

HelloKitty committed:
Address enum codestyle

jackpoz commented:
could you merge 3.3.5 branch into this PR ?

HelloKitty commented:
@jackpoz Last time I tried to do that I messed it up and lost some of my sanity.

jackpoz commented:
didn't you do a rebase last time ? I'm asking for a merge, that should be easier
- checkout 335
- pull TC 335
- switch to the PR branch
- merge 335 into the PR branch

HelloKitty committed:
Merge branch 'TrinityCore/3.3.5' into 3.3.5-creaturedbc

jackpoz committed:
Update DBCStructure.h

jackpoz committed:
Update DBCfmt.h

jackpoz commented:
Thanks for the PR :)
